compiler/rustc_codegen_llvm/src/gotoc/logging.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+use crate::btree_map;
+use rustc_serialize::json::*;
+use std::env;
+use std::fs::OpenOptions;
+use std::io::{self, prelude::*};
+
+pub use crate::{rmc_log, rmc_warn};
+
+struct LogLine {
+    log_type: LogType,
+    message: String,
+}
+
+impl ToJson for LogLine {
+    fn to_json(&self) -> Json {
+        let output = btree_map![
+            ("log_type".to_string(), self.log_type.to_json()),
+            ("message".to_string(), self.message.to_json())
+        ];
+
+        Json::Object(output)
+    }
+}
+
+enum LogType {
+    Log,
+    Warning(WarningType),
+}
+
+impl ToJson for LogType {
+    fn to_json(&self) -> Json {
+        match self {
+            LogType::Log => {
+                let output = btree_map![("type".to_string(), "LOG".to_json())];
+                Json::Object(output)
+            }
+            LogType::Warning(warning_type) => {
+                let output = btree_map![
+                    ("type".to_string(), "WARNING".to_json()),
+                    ("warning_kind".to_string(), warning_type.to_json())
+                ];
+                Json::Object(output)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum WarningType {
+    Unsupported,
+    Other,
+}
+
+impl ToJson for WarningType {
+    fn to_json(&self) -> Json {
+        format!("{:?}", self).to_json()
+    }
+}
+
+fn write_to_log_file(log_line: LogLine) -> io::Result<()> {
+    let line = format!("{}\n", log_line.to_json().to_string());
+    match env::var("RMC_LOG_FILE") {
+        Err(_) => Ok(()),
+        Ok(path) => {
+            OpenOptions::new().append(true).create(true).open(path)?.write_all(line.as_bytes())
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! rmc_log {
+    ($( $parts:expr ),*) => {
+        use crate::gotoc::logging::write_log;
+        let message = format!($( $parts, )*);
+        write_log(message).unwrap();
+    }
+}
+
+pub fn write_log(message: String) -> io::Result<()> {
```

Comment:
Should this be private?

Reply:
Since it's exposed in the macro, it can't be private (I don't think?).